ENH: Treat warnings as errors

jhlegarreta opened the pull request:
Treat warnings as errrors.

effigies commented:
I would suggest not adding -Werror for the min runs. Warnings are expected with old versions.

jhlegarreta commented:
Tried different things here, ended up resorting to https://github.com/nipreps/nifreeze/pull/40 solution.

As for
> I would suggest not adding -Werror for the min runs. Warnings are expected with old versions.

If you have a suggestion to avoid applying the option for them, I would be happy to add it.

effigies commented:
You could do something like this in tox.ini:

```cfg
command =
    pytest ... \
    !min:  -Werror \
      {posargs:-n auto}
```

effigies commented on tox.ini:
```diff
@@ -67,7 +67,7 @@ uv_resolution =
 
 commands =
-  pytest --durations=20 --durations-min=1.0 --cov-report term-missing {posargs:-n auto}
+  pytest --durations=20 --durations-min=1.0 --cov-report term-missing !min: -Werror {posargs:-n auto}
```
```suggestion
  pytest --durations=20 --durations-min=1.0 --cov-report term-missing \
  !min: -Werror \
  {posargs:-n auto}
```

effigies commented on tox.ini:
```diff
@@ -67,7 +67,9 @@ uv_resolution =
 commands =
-  pytest --durations=20 --durations-min=1.0 --cov-report term-missing {posargs:-n auto}
+  pytest --durations=20 --durations-min=1.0 --cov-report term-missing \
+  min,pre: -Wignore \
```
This warning setting takes precedence over those in `tool.pytest.ini_options`. For `min`, we expect a lot of warnings that the older versions of the libraries we're testing of course cannot resolve. For `pre`, we expect many `PendingDeprecationWarning`s and `DeprecationWarning`s before our upstreams even have time to address.

oesteban commented:
Added to [discussion for next TechMon](https://hackmd.io/tfD8wWnCQEuifdLpckO3ng?view=&stext=1056%3A36%3A0%3A1736844687%3AJ4mzO5)

jhlegarreta commented:
@effigies 8744db8 is the best solution I could come up with after many trial and error attempts; the thing is working locally for me. CIs show the right command:
https://github.com/nipreps/nireports/actions/runs/13276532890/job/37067053254?pr=157#step:14:40
```
py310-latest: commands[0]> pytest --durations=20 --durations-min=1.0 --cov-report term-missing -Werror -n auto
```
and
https://github.com/nipreps/nireports/actions/runs/13276532890/job/37067054313?pr=157#step:14:40
```
py39-min: commands[0]> pytest --durations=20 --durations-min=1.0 --cov-report term-missing -n auto
```
but py310 is failing for another reason and stopping all other executions.

The issue with the e.g. solution in a73c68f seems to be that `{!min:-Werror}` is being treated as a string literal and making the pytest calls fail.

jhlegarreta commented:
The solution in 4de5fca stops all jobs whenever a warning is encountered. Ideally, we would like all CIs to run and finish independently and only then report warnings. This can be achieved if the solution in https://github.com/nipreps/nifreeze/pull/40 is adopted, so I'd vote for adopting it here.

BTW, the Python 3.9 min build is raising some warnings that only appear in that build:
https://github.com/nipreps/nireports/actions/runs/13577984850/job/37958440074#step:14:488
and some of them having been already addressed, e.g.
https://github.com/nipreps/nireports/actions/runs/13577984850/job/37958440074#step:14:514
addressed in
https://github.com/nipreps/nireports/pull/156

Similarly, Python 3.11 is treating an already fixed warning as an error in this PR:
https://github.com/nipreps/nireports/actions/runs/13578088021/job/37958736735?pr=157#step:14:331
which was addressed in.
https://github.com/nipreps/nireports/pull/166

These two things are weird to me.

jhlegarreta commented:
So compared to 4de5fca, ebc9d39 allows all jobs to complete, instead of cancelling the rest of the jobs when one of them fails, so it looks a better solution to me. Now,
- Python 3.9 min is the only build containing warnings, but warnings are not raising an error. The proposed solution was working for [`NiFreeze`](https://github.com/nipreps/nifreeze/pull/40#issuecomment-2558646923), so not sure what's wrong in here.
- As noted in https://github.com/nipreps/nireports/pull/157#issuecomment-2689397229, some of the warnings in Python 3.9 min had already been addressed. They seem to still have been present in the Python 3.9 min builds.

A review from somebody else would be greatly appreciated.

jhlegarreta commented:
@effigies Thanks for recovering this. I rebased on `main` and removed what I had added to test that the warnings are considered as errors:

<img width="799" height="91" alt="nireports_warnings_as_errors_code" src="https://github.com/user-attachments/assets/4c65124d-4fb0-45b4-bf5b-53be125045c2" />

and it is doing its job as we've seen:

<img width="987" height="208" alt="nireports_warnings_as_errors_build" src="https://github.com/user-attachments/assets/f80bdb78-2e63-4f26-87e1-212c5cbd7595" />

So it should be good to go now.
